controllers/persistentvolumeclaim_controller.go

```diff
@@ -2,35 +2,194 @@ package controllers
 
 import (
 	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
 
 	"github.com/go-logr/logr"
 	corev1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
 	"k8s.io/apimachinery/pkg/runtime"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
 )
 
+const resizeEnableIndexKey = ".metadata.annotations[resize.topolvm.io/enabled]"
+const storageClassNameIndexKey = ".spec.storageClassName"
+
 // PersistentVolumeClaimReconciler reconciles a PersistentVolumeClaim object
 type PersistentVolumeClaimReconciler struct {
 	client.Client
 	Log    logr.Logger
 	Scheme *runtime.Scheme
+	MetricsClient
 }
 
-// +kubebuilder:rbac:groups=core,resources=persistentvolumeclaims,verbs=get;list;watch;create;update;patch;delete
-// +kubebuilder:rbac:groups=core,resources=persistentvolumeclaims/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=core,resources=persistentvolumeclaims,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups=storage.k8s.io,resources=storageclasses,verbs=get;list;watch
 
 func (r *PersistentVolumeClaimReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	_ = context.Background()
-	_ = r.Log.WithValues("persistentvolumeclaim", req.NamespacedName)
+	ctx := context.Background()
+	log := r.Log.WithValues("persistentvolumeclaim", req.NamespacedName)
 
 	// your logic here
+	var pvc corev1.PersistentVolumeClaim
+	err := r.Get(ctx, req.NamespacedName, &pvc)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	namespace, name := req.NamespacedName.Namespace, req.NamespacedName.Name
+	vs, err := r.MetricsClient.GetMetrics(ctx, namespace, name)
+	if err == errNotFound {
+		return ctrl.Result{
+			Requeue:      true,
+			RequeueAfter: 30 * time.Second,
+		}, nil
+	} else if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	threshold, err := convertSizeInBytes(pvc.Annotations[ResizeThresholdAnnotation], vs.CapacityBytes, DefaultThreshold)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	increase, err := convertSizeInBytes(pvc.Annotations[ResizeIncreaseAnnotation], pvc.Spec.Resources.Limits.Storage().Value(), DefaultIncrease)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	preCap, exist := pvc.Annotations[PreviousCapacityBytesAnnotation]
+	if exist {
+		preCapInt64, err := strconv.ParseInt(preCap, 10, 64)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		if preCapInt64 != vs.CapacityBytes {
+			delete(pvc.Annotations, PreviousCapacityBytesAnnotation)
+			err = r.Client.Update(ctx, &pvc)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+			log.Info("resize success!", "namespace", req.Namespace, "name", req.Name, "capacity", vs.CapacityBytes)
+			return ctrl.Result{}, nil
+		}
```

Comment:
I think, in this case, the resize should be done just the same as when there is no annotation.